pkg/controller/config/config_controller.go

```diff
@@ -273,20 +285,23 @@ func (r *ReconcileConfig) Reconcile(ctx context.Context, request reconcile.Reque
 	if r.needsReplay == nil {
 		r.needsReplay = r.watched.Intersection(newSyncOnly)
 	}
+
+	// Wipe all data to avoid stale state if needed. Happens once per watch-set-change.
+	if err := r.wipeCacheIfNeeded(ctx); err != nil {
+		return reconcile.Result{}, fmt.Errorf("wiping opa data cache: %w", err)
+	}
+
+	r.watchedMtx.Lock()
```

Comment:
GitHub wont let me comment on the old thread (against the removed TODOs that were here).

The race condition is between the config controller and the audit controller:

The main concern is that audit-from-cache cannot `List()` resources that are not currently being watched, since that could lead to dangling watches.

Audit correctly locks the watchset using DoForEach (which is clever), but the missing piece is that the config controller doesn't need to hold the watchset lock in order to call `r.watcher.ReplaceWatch()`, which is what could clear the cache.

Reply:
Does the `r.watchedMtx` not do this? What should be done instead?

Reply:
It'd have to be the same mutex that DoForEach() acquires, so that audit's call to DoForEach() blocks the watcher and vice-versa.

Reply:
Done and added race condition tests.